### DISCUSSION TITLE
Intent to move Apple projects to their own repository

### DISCUSSION THREAD

**Discussion by indygreg** [indygreg/PyOxidizer#643] [General]
This project has evolved into a dumping ground of random open source projects that I've started.

While I'm a huge proponent of monorepos, there are limits to their usefulness, especially when there isn't more formal backing of a monorepo by a larger, more formal entity (like a company).

There has been growing tension among the projects in this repository. A lot of people care about the Python projects. A lot of people care about the Apple projects (especially `apple-codesign`). And a lot of people care about other random projects in this repository.

While I'm personally in favor of the monorepo because it solves a lot of technical complexities (such as being able to make atomic commits touching multiple projects), I think the disjoint projects and target audiences is creating more problems than it is solving. It is constraining me from doing things that I think are healthy for projects' futures (such as adding maintainers). **Therefore, I've decided to start breaking up this monorepository.**

Initially, I'll be splitting out all the `apple-*` Rust crates into a new repository. This is tentatively https://github.com/indygreg/apple-platform-rs. However, I may promote this repository to its own GitHub organization in due time. Once I'm satisfied with the state of this new repository, I'll start transferring relevant issues to the new repository.

There are some opportunities for additional crates to be split into their own repository (notably the cryptography and Linux packaging crates). I haven't made a decision about the future of these crates at this time. However, in my mind the writing is on the wall.

**Comment by indygreg** [comment]
As feared (but not articulated in my original comment), some diamond dependency problems arose from splitting out the Apple crates into their own repository. e.g. crates relying on tugger-file-manifest, x509-certificate, and more were pulling in a published version from the new repo and a development, filesystem relative version from the PyOxidizer repo. As soon as these streams crossed, compiler errors ensued.

So, I'm now in the process of moving crates to their own repositories:

* https://github.com/indygreg/simple-file-manifest-rs/ (for what was formerly known as `tugger-file-manifest`)
* https://github.com/indygreg/cryptography-rs (for `cryptographic-message-syntax`, `pgp-cleartext`, and `x509-certificate`.

More crates may need to be moved. I'll know once I've shaved more of the yak.

**Reply by roblabla** [reply]
Unrelated, but did you think of perhaps upstreaming x509-certificate to the RustCrypto? They have the [x509-cert](https://docs.rs/x509-cert/latest/x509_cert/) crate, but it is extremely incomplete (doesn't support verification/validation, none of the higher level convenience for digest/signature/key algos). Maybe it would be useful to unify them?

**Reply by indygreg** [reply]
Yeah, I should probably have a convo with the RustCrypto folks. I'm not too keen on maintaining x509-certificate, as there are plenty of other crates we could use. If I had to start things over, I'm pretty sure I wouldn't need to create x509-certificate in the first place, as the maturity of the other X.509 crates has matured substantially.

**Comment by indygreg** [comment]
I just pushed commits cf8c4fe38501f8fa79027c51b35da3b736b72432..e744d0b6d2fb0848d39f9d4590f04793b126818a to the `main` branch that move a handful of projects to different repositories:

* `apple-bundles` -> https://github.com/indygreg/apple-platform-rs
* `apple-codesign` -> https://github.com/indygreg/apple-platform-rs
* `apple-flat-package` -> https://github.com/indygreg/apple-platform-rs
* `apple-xar` -> https://github.com/indygreg/apple-platform-rs
* `cryptographic-message-syntax` -> https://github.com/indygreg/cryptography-rs
* `pgp-cleartext` -> https://github.com/indygreg/cryptography-rs
* `tugger-file-manifest` -> https://github.com/indygreg/simple-file-manifest-rs/ and renamed to `simple-file-manifest`
* `x509-certificate` -> https://github.com/indygreg/cryptography-rs

218 files changed, 138 insertions(+), 47012 deletions(-)

GitHub issues for affected crates should have been moved as well.

All 3 new repos were created as clones of this repo so full Git history with original commit SHA-1s is preserved. I thought this would be better than rewriting Git history and destroying SHA-1 references. It makes the repos a bit larger than they should be. But the repo isn't that large, so I think it is fine.
